### DISCUSSION TITLE
How to modify some FooView's ViewState using value from FooEnvironment?

### DISCUSSION THREAD

**Discussion by acosmicflamingo** [pointfreeco/swift-composable-architecture#1263] [Q&A] [answered]
Let's say you have a FooViewController and a FooEnvironment struct with a closure that retrieves device orientation (useful because UIDevice.current.orientation returns .unknown in SwiftUI Previews, and seems like the kind of thing that Environment is meant to handle):
```
class FooViewController: UIViewController {
  struct ViewState {
    var isPortrait: Bool
  }
}

struct FooEnvironment {
  var getDeviceOrientation(): () -> UIDeviceOrientation
}
```

What I can't get my head around is how I would be able to modify FooView.ViewState.isPortrait because it is outside the scope of FooReducer (remember that I am keeping all view-specific logic in FooViewController.ViewState instead of FooState). Could I just put it in FooState? Sure, but doesn't that just pollute FooState? Am I missing something here? Thanks for any input you may have!

**Comment by tgrapperon** [comment]
The obvious path is your `State` indeed, but if the orientation doesn't play any role in your business logic, you can also use SwiftUI's `Environment` to pass the information around, and use it where it is needed, without being hosted in `ViewState`. There is no `Environment` in UIKit, but this particular dependency can also be retrieved on the spot from `UIDevice.current`. You probably can keep it uncontrolled if it doesn't play a role with your business logic, as this is then something you're not going to test anyway. If this is something that plays a role with your model, for example if it condition the existence of a view/feature, then it should probably go into `State` like a bona fide property.

**Reply by acosmicflamingo** [reply]
A bit related to what I asked a few months ago, can you think of a clean way of doing what isowords does with deviceState: https://github.com/pointfreeco/isowords/blob/main/Sources/Styleguide/DeviceState.swift#L47?

It's quite complicated in UIKit since @Environment does not exist, the closest-behaving feature I could use is @Dependency but that should only be used in reducers, and adding a cancellable set to a feature's State to allow subscribing to publishers seems to break the core principles of TCA.

It seems like I'm going to have to create some sort of wrapper class that sends notifications to AsyncStreams like this: https://github.com/pointfreeco/swift-composable-architecture/discussions/1743  ...unless you have any better suggestions :D

**Reply by tgrapperon** [reply]
With UIKit, you're almost bound to use publishers and subscriptions. You can easily generate a publisher from a view store using `viewStore.publisher.someValue`. But if the data doesn't play a role with your model (in other words, you won't test it), you can probably use Combine without thinking about it twice.

Depending on your requirements, you may need to observe the value from a `UIViewController` or `UIView` (like `UITraitCollection`). In this case, you can directly bubble up the value to TCA through the `ViewStore` (if this value affects your business logic). You can sum up the case like this:

1. Needs to be observed from UIKit, affects business logic -> send actions through a `ViewStore` to be processed by TCA.
2. Needs to be observed from UIKit, doesn't affect the business logic -> Subscribe with Combine at `viewWillAppear` for example, and directly update your `UIView[Controller]` without involving TCA (or use any alternative observation mechanism you may prefer)
3. Doesn't need to be observed from UIKit, affects business logic -> make a dependency, update the `State` and subscribe through `ViewStore.publisher.theValue` at `viewWillAppear`.
4. Doesn't need to be observed from UIKit, doesn't affect the business logic -> Same as "2", vanilla observation of you choice.

I hope it helps. I must admit I'm not really seeing any problem here, as it seems that you have total control over what's going on. It is perfectly fine to use Combine on vanilla UIKit (only modelling Combine-based `Effect`s in TCA is discouraged), but in any case, you have many workarounds. There are also many callbacks functions in `UIView/UIViewController` that you can override to directly send action to through the `ViewStore` (provided your feature needs it), so there are many cases where you won't need Combine at all.

**Reply by acosmicflamingo** [reply]
Thanks; it certainly does help to see all my options laid out for me! I think what makes it a "bit" more complicated is that I essentially want to pass a particular UIView a store that does not need to be in ParentStore's State since it'll never be used in its business logic, but should have a global behavior.

Let's say I want to create a UIView `SplitScreenView` that presents two subviews, and the subview layout is dependent on the device orientation:
```swift
public class SplitScreenView: UIView {
  var store: StoreOf<DeviceReducer>
  var viewStore: ViewStore<ViewStore, DeviceReducer.Action>
  
  public struct ViewStore {
    var orientation: UIDeviceOrientation

    public init(state: DeviceReducer.State) {
      self.orientation = state.currentOrientation
    }
  }

  /* Add Combine logic that subscribes to viewStore.publisher.orientation to change layout accordingly */
}
```
So far it's simple enough since it goes with what you said in #1. But there are 2 issues that now come up that I've been stuck on for months now:

1. What do you pass to SplitScreenView's store parameter?
2. How does `DeviceReducer.State.orientation` know when to update?

The two potential solutions for challenge #1 are:
```swift
public class SettingsViewController: UIViewController {
  public override func loadView() {
    self.view = SplitScreenView(store: ???)

    // Potential Solution #1
    //   - Technically solves problem but because `Scope` is not being used, then that means
    //     two views will have their own copy of DeviceReducer
    self.view = SplitScreenView(
      store: .init(
        initialState: .init()
        reducer: DeviceReducer()
      )
    )

    // Potential Solution #2
    //   - Solves copy problem but doesn't make sense because SettingsViewController
    //     should not have anything related to DeviceReducer because DeviceReducer is only
    //     useful within SplitScreenView; nothing outside of that domain should care about orientation
    self.view = SplitScreenView(
      store: self.store.scope(state: \.deviceState, action: Action.deviceReducer)
    )
  }
}
```

The use-case for challenge #2 would be this:
```swift
public struct DeviceReducer: ReducerProtocol {
  public struct State: Equatable {
    public var currentOrientation: UIDeviceOrientation
  }

  public enum Action: Equatable {
    case updateOrientation
  }

  @Dependency(\.idiom) var idiom
  @Dependency(\.orientation) var getCurrentOrientation  // { UIDevice.current.orientation } 

  public func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
    switch action {
    // how can this be called automatically?
    case .updateOrientation:
      state.orientation = getCurrentOrientation()
      return .none
    }
  }

  // Potential Solution #1
  // - Seems 1000% hacky and against TCA core principles
  var orientationPublisher = NotificationCenter.default.publisher(for: UIDevice.orientationDidChangeNotification)
  var cancellables = Set<AnyCancellable>()

  public init() {
    orientationPublisher
      .sink { _ in
        // Some code that somehow makes it possible to run .send(.updateOrientation)
      }.store(in: &cancellables)
  }
}
```

I keep see-sawing between treating DeviceState as a dependency and treating it as a reducer, and then kicking the can down the road because neither approaches seem to work cleanly for me. Perhaps a fresh pair of eyes may see something I'm totally missing :D is there anything you think I'm missing that would make implementing this kind of functionality a lot easier for me?

**Reply by tgrapperon** [reply]
The real question is "Does TCA need to know the device orientation to reduce action?". The device orientation is relatively straightforward to access. You can for example build the `AsyncStream<UIDeviceOrientation>`:
```swift
func deviceOrientationStream() -> AsyncStream<UIDeviceOrientation> {
  AsyncStream(
    UIDeviceOrientation.self,
    bufferingPolicy: .bufferingNewest(1)
  ) { continuation in
    let task = Task { @MainActor in
	  continuation.yield(UIDevice.current.orientation)
      for await _ in  NotificationCenter.default.notifications(named: UIDevice.orientationDidChangeNotification) {
        continuation.yield(UIDevice.current.orientation)
      }
    }
    continuation.onTermination = { _ in
      task.cancel()
    }
  }
}
```
and observe this stream. If you need to work with the orientation in TCA, you can follow the [long-living `Effect` SwiftUI case study](https://github.com/pointfreeco/swift-composable-architecture/blob/418e321acf4ab5318ab142e87f4a288b2ecb1b17/Examples/CaseStudies/SwiftUICaseStudies/02-Effects-LongLiving.swift) to observe this stream and update your state accordingly. You can then subscribe to `viewStore.publishers.deviceOrientation` in your `UIView/UIViewController` and perform whatever work you need to perform there. BTW, the subscription in the `Reducer` won't work, as theses values are regenerated for each action received, but I think the case-study shows the correct implementation of what you're trying to achieve there.

Of course, if TCA doesn't need to be involved, you can directly observe this stream from UIKit, and perform the same work as you would do if the values were coming from TCA.

Your `UIView/UIViewController` should accept a `Store` which domain corresponds to the domain that your UIKit component needs to work, like you would do with SwiftUI views. If you want to share the current `UIDeviceOrientation` value between features and be sure that the value is the same for everyone, you should observe it from a common parent and inject it into the states of the children (check the shared state case study if needed). The `ViewStore` of each child will publish changes automatically, even if their "associated" reducer is not observing the stream itself.

**Reply by acosmicflamingo** [reply]
Huh, that's a great idea!!! I haven't encountered any use-cases where I needed to get familiar with AsyncStream, but it appears that I have finally come across a scenario where it seems like the perfect solution to what I need. Thank you very much for your very detailed, helpful answers :D

**Reply by acosmicflamingo** [reply]
> ```swift
> func deviceOrientationStream() -> AsyncStream<UIDeviceOrientation> {
>   AsyncStream(
>     UIDeviceOrientation.self,
>     bufferingPolicy: .bufferingNewest(1)
>   ) { continuation in
>     let task = Task { @MainActor in
> 	  continuation.yield(UIDevice.current.orientation)
>       for await _ in  NotificationCenter.default.notifications(named: UIDevice.orientationDidChangeNotification) {
>         continuation.yield(UIDevice.current.orientation)
>       }
>     }
>     continuation.onTermination = { _ in
>       task.cancel()
>     }
>   }
> }
> ```

This works like a charm!!! Heck yeah :D :D :D

**Comment by acosmicflamingo** [comment, marked as answer]
This answer is what finally addressed the issue I was facing :) https://github.com/pointfreeco/swift-composable-architecture/discussions/1263#discussioncomment-4412801 all thanks to the helpful @tgrapperon
